### PR TITLE
Material generators from Afterlight

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Today in: i port moony's code for her.

Plasma gens and Uranium gens are now supplied with their corresponding materials. Tweaking the target power can make them last longer or shorter. they also work with upgrades (yay).

**Media**

image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:Moony & EmoGarbage404
- add: Plasma and Uranium generators now are powered by their corresponding materials.
- tweak: Increased output of Plasma and Uranium generators.

### CODE REVIEW

**Review by moonheart08 — APPROVED** [review]
nepotism

File: Content.Client/Power/Generator/GeneratorWindow.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+[GenerateTypedNameReferences]
+public sealed partial class GeneratorWindow : FancyWindow
+{
+    private SolidFuelGeneratorComponentBuiState? _lastState;
+
+    public GeneratorWindow(SolidFuelGeneratorBoundUserInterface bui, EntityUid vis)
+    {
+        RobustXamlLoader.Load(this);
+        //IoCManager.InjectDependencies(this);
```
Comment (AJCM-git) [inline Content.Client/Power/Generator/GeneratorWindow.xaml.cs:16]: a

**Comment by LankLTE** [comment]
> ## About the PR
> Today in: i port moony's code for her.
> 
> Plasma gens and Uranium gens are now supplied with their corresponding materials. Tweaking the target power can make them last longer or shorter. they also work with upgrades (yay).
> 
> **Media**
> 
> image
> 
> * [x]  I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase
> 
> **Changelog**
> 
> 🆑Moony & EmoGarbage404
> 
> * add: Plasma and Uranium generators now are powered by their corresponding materials.
> * tweak: Increased output of Plasma and Uranium generators.

Aren’t these used in mapping for when you need something to be infinitely powered (like shuttles n stuff). I feel like this’ll mess up quite a bit of pre-existing stuff

**Comment by LankLTE** [comment]
> I changed it for the cargo shuttle. If anything else uses it we'll find out soon enough.

I might be stupid but I thought literally every shuttle (salvage, evac, ERT) uses them

File: Content.Server/Power/Generator/ChemicalFuelGeneratorDirectSourceComponent.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+/// </summary>
+[RegisterComponent]
+public sealed class ChemicalFuelGeneratorDirectSourceComponent : Component
+{
+    /// <summary>
+    /// The solution to pull fuel material from.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("solution")]
+    public string Solution { get; set; } = "default";
```
Comment (AJCM-git) [inline Content.Server/Power/Generator/ChemicalFuelGeneratorDirectSourceComponent.cs:14]: getter/setter (also is default a valid value here for solutions or should this field be nullable)

File: Content.Server/Power/Generator/GasPowerProviderComponent.cs
Comment (AJCM-git) [inline Content.Server/Power/Generator/GasPowerProviderComponent.cs:1]: newline between fields

File: Content.Server/Power/Generator/GasPowerProviderComponent.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+using Content.Shared.Atmos;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom;
+
+namespace Content.Server.Power.Generator;
+
+/// <summary>
+/// This is used for providing gas power to machinery.
+/// </summary>
+[RegisterComponent]
+public sealed class GasPowerProviderComponent : Component
```
Comment (AJCM-git) [inline Content.Server/Power/Generator/GasPowerProviderComponent.cs:11]: acess

File: Content.Server/Power/Generator/GasPowerProviderSystem.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                    if (mix is not null)
+                        _atmosphereSystem.Merge(res, mix);
+                }
+
+                SetPowered(uid, component, true);
+            }
+            else
+            {
+                SetPowered(uid, component, false);
+            }
```
Comment (AJCM-git) [inline Content.Server/Power/Generator/GasPowerProviderSystem.cs:75]: maybe some comments (or moving conditions to named variables) to make this more readable?

File: Content.Server/Power/Generator/GeneratorSystem.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        SubscribeLocalEvent<ChemicalFuelGeneratorAdapterComponent, InteractUsingEvent>(OnChemicalFuelAdapterInteractUsing);
+        SubscribeLocalEvent<FuelGeneratorComponent, SetTargetPowerMessage>(OnTargetPowerSet);
+    }
+
+    private void OnChemicalFuelAdapterInteractUsing(EntityUid uid, ChemicalFuelGeneratorAdapterComponent component, InteractUsingEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!TryComp(args.Used, out SolutionContainerManagerComponent? solutions) || !TryComp(uid, out FuelGeneratorComponent? generator))
```
Comment (AJCM-git) [inline Content.Server/Power/Generator/GeneratorSystem.cs:30]: use the generic TryComp<T>() (same for others in the file)

File: Content.Server/Power/Generator/GeneratorSystem.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        component.TargetPower = Math.Clamp(args.TargetPower, 0, component.MaxTargetPower / 1000) * 1000;
+    }
+
+    private void OnSolidFuelAdapterInteractUsing(EntityUid uid, SolidFuelGeneratorAdapterComponent component, InteractUsingEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!TryComp(args.Used, out PhysicalCompositionComponent? mat) || !TryComp(args.Used, out StackComponent? stack) || !TryComp(uid, out FuelGeneratorComponent? generator))
```
Comment (AJCM-git) [inline Content.Server/Power/Generator/GeneratorSystem.cs:75]: separate conditions with one line

File: Content.Shared/Power/Generator/FuelGeneratorComponent.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
+﻿using Robust.Shared.Serialization;
+
+namespace Content.Shared.Power.Generator;
+
+/// <summary>
+/// This is used for generators that run off some kind of fuel.
+/// </summary>
+[RegisterComponent]
+public sealed class FuelGeneratorComponent : Component
```
Comment (AJCM-git) [inline Content.Shared/Power/Generator/FuelGeneratorComponent.cs:9]: acess, separate fields with one line

**Comment by EmoGarbage404** [comment]
replaced all the generators in shuttles with 15kW basic generators

**Review by AJCM-git — CHANGES_REQUESTED** [review]
mostly details

**Comment by Arendian** [comment]
Terminal area, Nukie planet and centcomm also use them for power

**Comment by EmoGarbage404** [comment]
replaced uranium/plasma gens in remaining maps.

> Is uranum common enough for it to be fuel for a generator?

yeah salvage is just lazy

> Almost no department can currently run on one uranium generator(15KW), so one plasma generator producing 20KW at full power may still not be able to power a singular department by itself.

well uranium gen is buffed to like 30k plus you can run them at higher output for less efficiency. You could probably run a single department on a uranium/plasma gen if you get thrifty with your power consumption.

**Comment by Anzarot121** [comment]
Core use two uranium gens to create sungilarity.

**Comment by VektorZ1** [comment]
It would be cool to have a state where a small amount of energy is produced, at no cost, or minimal cost (0.5-1 ingots per hour or something like that.) This would be useful for creating autonomous shuttles and other systems.
For balance, as an option it would require full refuelling (or >80% in the spend option), and not generate much energy.

**Comment by deltanedas** [comment]
we still have apu spam, for now

File: Content.Shared/Power/Generator/FuelGeneratorComponent.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
+﻿using Robust.Shared.Serialization;
+
+namespace Content.Shared.Power.Generator;
+
+/// <summary>
+/// This is used for generators that run off some kind of fuel.
+/// </summary>
+[RegisterComponent, Access(typeof(SharedGeneratorSystem))]
```
Comment (metalgearsloth) [inline Content.Shared/Power/Generator/FuelGeneratorComponent.cs:8]: NetworkedComponent

File: Content.Server/Power/Generator/GeneratorSystem.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        generator.RemainingFuel += stack.Count * component.Multiplier;
+        QueueDel(args.Used);
+        args.Handled = true;
+    }
+
+    public override void Update(float frameTime)
+    {
+        var query = EntityQueryEnumerator<FuelGeneratorComponent, PowerSupplierComponent, TransformComponent>();
+
+        var upgradeQuery = GetEntityQuery<UpgradePowerSupplierComponent>();
```
Comment (metalgearsloth) [inline Content.Server/Power/Generator/GeneratorSystem.cs:93]: Just dump it as a field on the class and get it on entity Initialize.

File: Content.Shared/Power/Generator/SharedGeneratorSystem.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+{
+    /// <summary>
+    /// Calculates the expected fuel efficiency based on the optimal and target power levels.
+    /// </summary>
+    /// <param name="targetPower">Target power level</param>
+    /// <param name="optimalPower">Optimal power level</param>
+    /// <returns>Expected fuel efficiency as a percentage</returns>
+    public static float CalcFuelEfficiency(float targetPower, float optimalPower)
+    {
+        return MathF.Pow(optimalPower / targetPower, 1.3f);
```
Comment (metalgearsloth) [inline Content.Shared/Power/Generator/SharedGeneratorSystem.cs:16]: Buh? Should this be a constant? Why 1.3?

File: Content.Server/Power/Generator/GasPowerProviderSystem.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+using Content.Server.Power.Components;
+using Content.Shared.Atmos;
+using Robust.Shared.Timing;
+
+namespace Content.Server.Power.Generator;
+
+/// <summary>
+/// This handles gas power providers, allowing devices to accept power in the form of plasma or high pressure gas.
+/// </summary>
+public sealed class GasPowerProviderSystem : EntitySystem
```
Comment (metalgearsloth) [inline Content.Server/Power/Generator/GasPowerProviderSystem.cs:15]: Isn't this receiving gas, not providing it?

File: Content.Server/Power/Generator/GeneratorSystem.cs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        if (args.Handled)
+            return;
+
+        if (!TryComp(args.Used, out PhysicalCompositionComponent? mat) ||
+            !TryComp(args.Used, out StackComponent? stack) ||
+            !TryComp(uid, out FuelGeneratorComponent? generator))
+            return;
+
+        if (!mat.MaterialComposition.Keys.ToList().Contains(component.FuelMaterial))
+            return;
```
Comment (metalgearsloth) [inline Content.Server/Power/Generator/GeneratorSystem.cs:82]: ContainsKey?

File: Content.Server/Power/Generator/GasPowerProviderSystem.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+    private void SetPowered(EntityUid uid, GasPowerProviderComponent comp, bool state)
+    {
+        if (state != comp.Powered)
+        {
+            comp.Powered = state;
+            var ev = new PowerChangedEvent(state, 0);
+            RaiseLocalEvent(uid, ref ev);
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Server/Power/Generator/GasPowerProviderSystem.cs:91]: Probably better off having an interface for IPowerReceivers and sharing it with ApcPowerReceiverComponent given they're similar in purpose.

File: Content.Server/Power/Generator/GasPowerProviderSystem.cs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        SubscribeLocalEvent<GasPowerProviderComponent, AtmosDeviceUpdateEvent>(OnDeviceUpdated);
+    }
+
+    private void OnDeviceUpdated(EntityUid uid, GasPowerProviderComponent component, AtmosDeviceUpdateEvent args)
+    {
+        if (component.LastProcess == TimeSpan.Zero)
+        {
+            component.LastProcess = _gameTiming.CurTime;
+            return; // Skip tick 0.
+        }
```
Comment (metalgearsloth) [inline Content.Server/Power/Generator/GasPowerProviderSystem.cs:33]: Set this on MapInit to current tick and you won't need this on every update.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Power/Generator/SharedGeneratorSystem.cs (hunk 10 above)
Comment (moonheart08) [reply Content.Shared/Power/Generator/SharedGeneratorSystem.cs:16]: It's a magic number I found during balancing.

File: Content.Server/Power/Generator/GasPowerProviderSystem.cs (hunk 13 above)
Comment (EmoGarbage404) [reply Content.Server/Power/Generator/GasPowerProviderSystem.cs]: I'm not really sure how to do this in a sane way without diving into how some of the internals of power code.

Comment (moonheart08) [reply Content.Server/Power/Generator/GasPowerProviderSystem.cs]: sloth is cring ignore them

**Review by AJCM-git — CHANGES_REQUESTED** [review]
add popups to the generator (when inserting and other stuff that may need player feedback)

**Review by AJCM-git — APPROVED** [review]

**Pull request merged**